sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -834,6 +834,140 @@ case class CollectObjectsToMap private(
+    collClass: Class[_]) extends Expression with NonSQLExpression {
+
+  override def nullable: Boolean = inputData.nullable
+
+  override def children: Seq[Expression] = lambdaFunction :: inputData :: Nil
+
+  override def eval(input: InternalRow): Any =
+    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
+
+  override def dataType: DataType = ObjectType(collClass)
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    // The data with PythonUserDefinedType are actually stored with the data type of its sqlType.
+    def inputDataType(dataType: DataType) = dataType match {
+      case p: PythonUserDefinedType => p.sqlType
+      case _ => dataType
+    }
+
+    val arrayType = inputDataType(inputData.dataType).asInstanceOf[ArrayType]
+    val loopValueJavaType = ctx.javaType(arrayType.elementType)
+    ctx.addMutableState("boolean", loopIsNull, "")
+    ctx.addMutableState(loopValueJavaType, loopValue, "")
+    val genFunction = lambdaFunction.genCode(ctx)
+
+    val genInputData = inputData.genCode(ctx)
+    val dataLength = ctx.freshName("dataLength")
+    val loopIndex = ctx.freshName("loopIndex")
+    val builderValue = ctx.freshName("builderValue")
+
+    val getLength = s"${genInputData.value}.numElements()"
+    val getLoopVar = ctx.getValue(genInputData.value, arrayType.elementType, loopIndex)
+
+    // Make a copy of the data if it's unsafe-backed
+    def makeCopyIfInstanceOf(clazz: Class[_ <: Any], value: String) =
+      s"$value instanceof ${clazz.getSimpleName}? $value.copy() : $value"
+    val genFunctionValue =
+      lambdaFunction.dataType match {
+        case StructType(_) => makeCopyIfInstanceOf(classOf[UnsafeRow], genFunction.value)
+        case ArrayType(_, _) => makeCopyIfInstanceOf(classOf[UnsafeArrayData], genFunction.value)
+        case MapType(_, _, _) => makeCopyIfInstanceOf(classOf[UnsafeMapData], genFunction.value)
+        case _ => genFunction.value
+      }
+
+    val loopNullCheck = s"$loopIsNull = ${genInputData.value}.isNullAt($loopIndex);"
+
+    val builderClass = classOf[Builder[_, _]].getName
+    val constructBuilder = s"""
+      $builderClass $builderValue = ${collClass.getName}$$.MODULE$$.newBuilder();
+      $builderValue.sizeHint($dataLength);
+    """
+
+    val appendToBuilder = s"""
+      if (${genFunction.isNull}) {
+        $builderValue.$$plus$$eq(null);
+      } else {
+        $builderValue.$$plus$$eq($genFunctionValue);
+      }
+     """
+    val getBuilderResult = s"${ev.value} = (${collClass.getName}) $builderValue.result();"
+
+    val code = s"""
+      ${genInputData.code}
+      ${ctx.javaType(dataType)} ${ev.value} = ${ctx.defaultValue(dataType)};
+
+      if (!${genInputData.isNull}) {
+        int $dataLength = $getLength;
+        $constructBuilder
+
+        int $loopIndex = 0;
+        while ($loopIndex < $dataLength) {
+          $loopValue = ($loopValueJavaType) ($getLoopVar);
+
+          $loopNullCheck
```

Comment:
I checked gen'd code;
```
scala> val ds = Seq(Seq(1), Seq(1, 2)).toDF("a").as[Set[Int]]
ds: org.apache.spark.sql.Dataset[Set[Int]] = [a: array<int>]

scala> ds.printSchema
root
 |-- a: array (nullable = true)
 |    |-- element: integer (containsNull = false)

scala> ds.filter(_.size > 3).debugCodegen
...
/* 056 */             int filter_loopIndex = 0;
/* 057 */             while (filter_loopIndex < filter_dataLength) {
/* 058 */               CollectObjectsToSet_loopValue6 = (int) (inputadapter_value.getInt(filter_loopIndex));
/* 059 */
/* 060 */               CollectObjectsToSet_loopIsNull6 = inputadapter_value.isNullAt(filter_loopIndex);
/* 061 */
/* 062 */               if (CollectObjectsToSet_loopIsNull6) {
/* 063 */                 filter_builderValue.$plus$eq(null);
/* 064 */               } else {
/* 065 */                 filter_builderValue.$plus$eq(CollectObjectsToSet_loopValue6);
/* 066 */               }
/* 067 */
/* 068 */               filter_loopIndex += 1;
/* 069 */             }
...
```
`containsNull = false` though, we need null check in this loop?